RFC: Faster methods for tuple comparison and hash

andyferris opened the pull request:
~~`isequal(::Tuple, ::Tuple)` previously used recursion, which got slow for tuples greater than size 14 (and if I understand correctly, this kind of recursion is not O(N) for very large tuples). Now it is fully unrolled generated code (when `if @generated`, otherwise it follows a `for` loop).~~

`isless(::Tuple, ::Tuple)` previously used a for loop in all cases, which is OK for the homogenous element types case but much slower otherwise. Now it is unrolled via recursion (for small tuples only). A similar approach is implemented for `<` and `hash`.

~~While some might see this as a bit "heavy", I think we need this kind of stuff for nice, generic, speedy operations on e.g. simple tables which are collections of `Tuples` and `NamedTuples`.~~ Recursion seems okay-ish at first but my limited understanding is that it won't always act O(N) for very large tuples.

Follow-up of #26025. What motivated both PR's was a slow `sortperm(::AbstractVector{<:NamedTuple})` and other operations you might want on sorted "table"-like data structures - @shashi, @JeffBezanson I'd be curious how the generic methods in `Base` for `sortperm` and `searchsorted` now compare to the (optimized) algorithms used in `IndexedTables`.

~~We might want to do something for `==` and `<` in the future for very similar reasons (FYI @nalimilan). Possibly also `hash`.~~ We might want to do something similar for `sum` and `prod`.

andyferris commented:
PS - getting this through bootstrap was a fun game. I ended up moving the `macro generated()` definition to essentials.

vtjnash commented:
> Recursion seems okay-ish

Yes, recursion is OK. Please just use recursion and let the inliner do its job, rather than using a generated function to intentionally disrupt inference.

> previously used a for loop in all cases, which is OK for the homogenous element types case but much slower otherwise

I'm slightly inclined to think we should use loops for `NTuple` and tail-recursion otherwise (instead of the current length cutoff), but that's a separate PR. For now, we should just continue to be consistent and use Any16.

andyferris commented:
> recursion is OK

The problem I'm observing is that recursion doesn't work for very large structures. The pattern I thought we were meant to follow was to have deeper calls have "simpler" input types than the caller, which I benchmark like so:

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.3980 (2018-02-14 21:32 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit f965fd48dd* (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> using BenchmarkTools

julia> @noinline f(x) = 2*x
f (generic function with 1 method)

julia> g(x::Tuple) = _g(x...)
g (generic function with 1 method)

julia> @inline _g(x, xs...) = (f(x), _g(xs...)...)
_g (generic function with 1 method)

julia> _g() = ()
_g (generic function with 2 methods)

julia> for i = 4:4:128
           print("n = $i"); @btime g($(ntuple(identity, i)))
       end
n = 4  8.533 ns (0 allocations: 0 bytes)
n = 8  14.656 ns (0 allocations: 0 bytes)
n = 12  21.116 ns (0 allocations: 0 bytes)
n = 16  15.713 μs (32 allocations: 1.61 KiB)
n = 20  21.732 μs (40 allocations: 2.34 KiB)
n = 24  28.437 μs (48 allocations: 3.19 KiB)
n = 28  35.054 μs (56 allocations: 4.16 KiB)
n = 32  43.577 μs (64 allocations: 5.38 KiB)
n = 36  52.118 μs (72 allocations: 6.75 KiB)
n = 40  61.252 μs (80 allocations: 8.25 KiB)
n = 44  72.412 μs (88 allocations: 9.88 KiB)
n = 48  83.694 μs (96 allocations: 11.59 KiB)
n = 52  95.704 μs (104 allocations: 13.41 KiB)
n = 56  108.494 μs (112 allocations: 15.41 KiB)
n = 60  122.551 μs (120 allocations: 17.59 KiB)
n = 64  137.166 μs (128 allocations: 19.78 KiB)
n = 68  152.243 μs (136 allocations: 22.03 KiB)
n = 72  168.191 μs (144 allocations: 24.44 KiB)
n = 76  185.418 μs (152 allocations: 27.03 KiB)
n = 80  203.052 μs (160 allocations: 29.72 KiB)
n = 84  221.566 μs (168 allocations: 32.53 KiB)
n = 88  240.411 μs (176 allocations: 35.47 KiB)
n = 92  261.323 μs (184 allocations: 38.56 KiB)
n = 96  281.338 μs (192 allocations: 41.81 KiB)
n = 100  303.394 μs (200 allocations: 45.22 KiB)
n = 104  326.301 μs (208 allocations: 48.78 KiB)
n = 108  350.029 μs (216 allocations: 52.34 KiB)
n = 112  372.901 μs (224 allocations: 56.13 KiB)
n = 116  397.670 μs (232 allocations: 60.13 KiB)
n = 120  422.614 μs (240 allocations: 64.13 KiB)
n = 124  448.633 μs (248 allocations: 68.28 KiB)
n = 128  474.895 μs (256 allocations: 72.59 KiB)
```

I observe that from `n=15` onwards (`n=6` if I don't add the `@inline`) the codegen gets pretty bad, and the worst part is that this is clearly not O(`n`) scaling... execution time scales worse than that.

With generated code I get:

```
julia> @generated function h(x::NTuple{n, Any}) where {n}
           exprs = [:(f(x[$j])) for j = 1:n]
           return Expr(:tuple, exprs...)
       end
h (generic function with 1 method)

julia> for i = 4:4:128
           print("n = $i"); @btime h($(ntuple(identity, i)))
       end
n = 4  8.531 ns (0 allocations: 0 bytes)
n = 8  17.038 ns (0 allocations: 0 bytes)
n = 12  23.498 ns (0 allocations: 0 bytes)
n = 16  30.639 ns (0 allocations: 0 bytes)
n = 20  42.884 ns (0 allocations: 0 bytes)
n = 24  51.043 ns (0 allocations: 0 bytes)
n = 28  59.202 ns (0 allocations: 0 bytes)
n = 32  66.723 ns (0 allocations: 0 bytes)
n = 36  75.878 ns (0 allocations: 0 bytes)
n = 40  82.681 ns (0 allocations: 0 bytes)
n = 44  92.214 ns (0 allocations: 0 bytes)
n = 48  99.686 ns (0 allocations: 0 bytes)
n = 52  108.192 ns (0 allocations: 0 bytes)
n = 56  116.362 ns (0 allocations: 0 bytes)
n = 60  129.325 ns (0 allocations: 0 bytes)
n = 64  132.353 ns (0 allocations: 0 bytes)
n = 68  141.189 ns (0 allocations: 0 bytes)
n = 72  147.994 ns (0 allocations: 0 bytes)
n = 76  157.525 ns (0 allocations: 0 bytes)
n = 80  165.022 ns (0 allocations: 0 bytes)
n = 84  361.134 ns (0 allocations: 0 bytes)
n = 88  181.684 ns (0 allocations: 0 bytes)
n = 92  189.528 ns (0 allocations: 0 bytes)
n = 96  201.737 ns (0 allocations: 0 bytes)
n = 100  207.234 ns (0 allocations: 0 bytes)
n = 104  214.383 ns (0 allocations: 0 bytes)
n = 108  224.278 ns (0 allocations: 0 bytes)
n = 112  231.288 ns (0 allocations: 0 bytes)
n = 116  252.849 ns (0 allocations: 0 bytes)
n = 120  248.185 ns (0 allocations: 0 bytes)
n = 124  256.808 ns (0 allocations: 0 bytes)
n = 128  265.082 ns (0 allocations: 0 bytes)
```

which is linear in `n`, is the same for `n=4..12`, and significantly faster otherwise.

The methods in this PR are really useful for comparing rows of tables. People will have heterogeneous types and more than 15 columns. 

Would we consider upping the recursion limits beyond 16?
 
> rather than using a generated function to intentionally disrupt inference

I would phrase it as "using a generated function to intentionally *help* inference" :laughing:  (I guess we could call the code in `base/compile/` something other than "inference", e.g. "the compiler", now?).

bramtayl commented:
Ref #23875

chethega commented:
I'd just like to remark that for` nt=ntuple(identity, 1000)`, `@time @code_native h(nt)` takes `1.387764 seconds (437.95 k allocations: 23.766 MiB, 0.88% gc time), looks absolutely disgusting (also look at `@code_llvm h(nt)`) and will totally destroy the instruction cache.

Isn't there a way to re-roll into a loop? Or are the limits for when to ouput a badly typed loop vs well-typed unrolled code outdated?

Also, the scaling for the recursive version `g` should be perfectly quadratic, which is borne out by a small experiment; for some reason compilation times for `h` are quadratic as well (which is irrelevant because the generated code for large tuples is too disgusting to emit) :
```
julia> for i in [100, 200, 400, 800, 1600, 3200]
       nt = ntuple(identity, i);
       println("n = $i:")
       @time g(nt);
       @btime g($nt);
       @time h(nt);
       @btime h($nt);
       println();
       end

n = 100:
  0.022626 seconds (6.78 k allocations: 466.973 KiB)
  282.446 ��s (200 allocations: 45.22 KiB)
  0.124854 seconds (78.59 k allocations: 4.605 MiB)
  256.108 ns (0 allocations: 0 bytes)

n = 200:
  0.006940 seconds (603 allocations: 193.497 KiB)
  1.034 ms (400 allocations: 171.03 KiB)
  0.113794 seconds (75.07 k allocations: 4.372 MiB)
  512.151 ns (0 allocations: 0 bytes)

n = 400:
  0.010643 seconds (48.56 k allocations: 1.404 MiB)
  4.359 ms (48359 allocations: 1.37 MiB)
  0.277382 seconds (158.00 k allocations: 9.022 MiB)
  1.066 ��s (0 allocations: 0 bytes)

n = 800:
  0.034620 seconds (290.72 k allocations: 11.400 MiB, 19.73% gc time)
  18.723 ms (290517 allocations: 11.35 MiB)
  0.867887 seconds (327.91 k allocations: 17.846 MiB)
  2.256 ��s (0 allocations: 0 bytes)

n = 1600:
  0.136879 seconds (1.26 M allocations: 55.801 MiB, 18.28% gc time)
  85.963 ms (1255717 allocations: 55.66 MiB)
  2.961486 seconds (668.94 k allocations: 36.566 MiB, 0.42% gc time)
  5.381 ��s (0 allocations: 0 bytes)

n = 3200:
  0.683033 seconds (5.11 M allocations: 232.473 MiB, 30.39% gc time)
  362.034 ms (5106117 allocations: 232.16 MiB)
 10.854160 seconds (1.35 M allocations: 71.776 MiB, 0.63% gc time)
  40.072 ��s (0 allocations: 0 bytes)
```

Maybe put in a cut-off at maybe 32-element tuples into the generated code?

andyferris commented:
Yes, I was thinking we want to go from unrolled to loop, not unrolled to recursive, as `n` grows here. 

We can do that with the generated approach or try work that into recursion with a different approach to deal with `Any16`. Any preference?

Then, there’s the fact that I feel a cutoff of 15 or 16 is too small to transition to completely slow behavior. 32 might suffice for a lot of table stuff (and hopefully the loop won’t be too slow in the large `n` case with heterogenous types anyway).

Jutho commented:
I recently changed `isequal` and `==` from always loops to the recursive behavior with loop fallback for `Any16`. 

@andyferris , I am a bit confused by "I observe that from n=15 onwards...". Isn't that just saying that the loop is still worse in this range of `n` than the recursive not behavior, rather than saying that the recursive behavior is not working?

Regarding limits / values for the transition behavior, I too would love a slightly higher value in the order 20 - 30, and with that the same larger value for `MAX_TUPLETYPE_LEN`. My motivation is different (I am thinking about arrays with `N=30` where every dimension has size `2`), but my request the same :-)

chethega commented:
Ok. So currently we have very good code for large homogeneous tuples, whereas inhomogeneous tuples become very bad for large N.

We should preserve the good thing for homogeneous tuples; and, ultimately, large inhomogeneous tuples are currently just bad. We can't even collect on them.

Maybe codegen should have a special case that stores them as `Vector{Any}`? 

```
julia> for i in [100, 200, 400, 800, 1600, 3200]
                     nt = ntuple(identity, i);
                     nt2= convert(Vector{Any}, 1:i)
                     nt2[1]=Float64(nt[1])
                     nt2=collect(nt2)
                     
                     println("n = $i:")
                     @btime isequal($nt, $nt)
                     @btime isequal($nt, $(nt2...,))
                     @btime isequal($(collect(nt)), $(collect(nt2)))
                     @btime collect($(nt2...,))
                     println();
                     end
n = 100:
  97.723 ns (0 allocations: 0 bytes)
  9.986 ��s (101 allocations: 87.52 KiB)
  4.238 ��s (0 allocations: 0 bytes)
  10.966 ��s (103 allocations: 88.41 KiB)

n = 200:
  184.917 ns (0 allocations: 0 bytes)
  37.730 ��s (201 allocations: 318.77 KiB)
  8.474 ��s (0 allocations: 0 bytes)
  39.424 ��s (203 allocations: 320.55 KiB)

n = 400:
  359.724 ns (0 allocations: 0 bytes)
  285.971 ��s (401 allocations: 1.25 MiB)
  16.745 ��s (0 allocations: 0 bytes)
  284.884 ��s (403 allocations: 1.25 MiB)

n = 800:
  708.135 ns (0 allocations: 0 bytes)
  779.198 ��s (1379 allocations: 4.94 MiB)
  35.109 ��s (289 allocations: 4.52 KiB)
  770.062 ��s (1670 allocations: 4.95 MiB)

n = 1600:
  1.412 ��s (0 allocations: 0 bytes)
  2.802 ms (3779 allocations: 19.66 MiB)
  72.027 ��s (1089 allocations: 17.02 KiB)
  2.837 ms (4870 allocations: 19.69 MiB)

n = 3200:
  2.808 ��s (0 allocations: 0 bytes)
  23.253 ms (8579 allocations: 78.40 MiB)
  146.337 ��s (2689 allocations: 42.02 KiB)
  23.914 ms (11271 allocations: 78.47 MiB)
```

Amusingly this also appears in the much simpler full-loop-version:

```
nt2t=(nt2...,); #tuple with 3199 int64 and one float64
function isequal2(a,b)
       for i=1:length(a)
          @inbounds isequal(a[i], b[i]) || return false
          end
          return true
          end
@btime isequal2($nt2t, $nt2t)
  51.323 ms (17158 allocations: 156.80 MiB)
true
```

If we look at the code, we see that this is not an issue of quadratic recursion; instead llvm pulls a copy in the inner loop! This is batshit crazy.

Maybe just document that large tuples suck in julia and hope that people stop using them?

```
@code_llvm isequal2(nt2t, nt2t)

; Function isequal2
; Location: REPL[733]:2
define i8 @julia_isequal2_69228({ double, many i64 } addrspace(11)* nocapture nonnull readonly dereferenceable(25600), { double, many 164} addrspace(11)* nocapture nonnull readonly dereferenceable(25600)) {
top:
  %2 = alloca [3 x %jl_value_t addrspace(10)*], align 8
  %gcframe9 = alloca [5 x %jl_value_t addrspace(10)*], align 8
  %gcframe9.sub = getelementptr inbounds [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 0
  %.sub = getelementptr inbounds [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %2, i64 0, i64 0
  %3 = getelementptr inbounds [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 1
  %4 = bitcast %jl_value_t addrspace(10)** %3 to i8*
  call void @llvm.memset.p0i8.i32(i8* %4, i8 0, i32 32, i32 8, i1 false)
  %thread_ptr = call i8* asm "movq %fs:0, $0", "=r"() #4
  %ptls_i8 = getelementptr i8, i8* %thread_ptr, i64 -10920
  %5 = bitcast [5 x %jl_value_t addrspace(10)*]* %gcframe9 to i64*
  store i64 6, i64* %5, align 8
  %6 = getelementptr [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 1
  %7 = bitcast i8* %ptls_i8 to i64*
  %8 = load i64, i64* %7, align 8
  %9 = bitcast %jl_value_t addrspace(10)** %6 to i64*
  store i64 %8, i64* %9, align 8
  %10 = bitcast i8* %ptls_i8 to %jl_value_t addrspace(10)***
  store %jl_value_t addrspace(10)** %gcframe9.sub, %jl_value_t addrspace(10)*** %10, align 8
; Location: REPL[733]:3
; Function getindex; {
; Location: tuple.jl:22
  %11 = bitcast { double, many i64} addrspace(11)* %0 to i8 addrspace(11)*
  %12 = getelementptr [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 3
  %13 = getelementptr [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 2
  %14 = getelementptr [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %2, i64 0, i64 1
  %15 = getelementptr [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %2, i64 0, i64 2
  %16 = bitcast { double, many i64 } addrspace(11)* %1 to i8 addrspace(11)*
  %17 = getelementptr [5 x %jl_value_t addrspace(10)*], [5 x %jl_value_t addrspace(10)*]* %gcframe9, i64 0, i64 4
;}
; Location: REPL[733]:2
  br label %L2

L2:                                               ; preds = %L57, %top
  %"#temp#.0" = phi i64 [ 1, %top ], [ %53, %L57 ]
  %18 = icmp eq i64 %"#temp#.0", 3201
  br i1 %18, label %L66, label %if

if:                                               ; preds = %L2
; Location: REPL[733]:3
; Function getindex; {
; Location: tuple.jl:22
  %19 = call noalias nonnull %jl_value_t addrspace(10)* @jl_gc_big_alloc(i8* %ptls_i8, i64 25608)
  %20 = bitcast %jl_value_t addrspace(10)* %19 to %jl_value_t addrspace(10)* addrspace(10)*
  %21 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)* addrspace(10)* %20, i64 -1
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139704619030112 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)* addrspace(10)* %21, align 8
  %22 = bitcast %jl_value_t addrspace(10)* %19 to i8 addrspace(10)*
  call void @llvm.memcpy.p10i8.p11i8.i64(i8 addrspace(10)* %22, i8 addrspace(11)* %11, i64 25600, i32 8, i1 false)
  store %jl_value_t addrspace(10)* %19, %jl_value_t addrspace(10)** %12, align 8
  %23 = call %jl_value_t addrspace(10)* @jl_box_int64(i64 signext %"#temp#.0")
  store %jl_value_t addrspace(10)* %23, %jl_value_t addrspace(10)** %13, align 8
  store %jl_value_t addrspace(10)* %19, %jl_value_t addrspace(10)** %.sub, align 8
  store %jl_value_t addrspace(10)* %23, %jl_value_t addrspace(10)** %14, align 8
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139705131220400 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %15, align 8
  %24 = call nonnull %jl_value_t addrspace(10)* @jl_f_getfield(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %.sub, i32 3)
  store %jl_value_t addrspace(10)* %24, %jl_value_t addrspace(10)** %13, align 8
  %25 = call noalias nonnull %jl_value_t addrspace(10)* @jl_gc_big_alloc(i8* %ptls_i8, i64 25608)
  %26 = bitcast %jl_value_t addrspace(10)* %25 to %jl_value_t addrspace(10)* addrspace(10)*
  %27 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)* addrspace(10)* %26, i64 -1
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139704619030112 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)* addrspace(10)* %27, align 8
  %28 = bitcast %jl_value_t addrspace(10)* %25 to i8 addrspace(10)*
  call void @llvm.memcpy.p10i8.p11i8.i64(i8 addrspace(10)* %28, i8 addrspace(11)* %16, i64 25600, i32 8, i1 false)
  store %jl_value_t addrspace(10)* %25, %jl_value_t addrspace(10)** %17, align 8
  %29 = call %jl_value_t addrspace(10)* @jl_box_int64(i64 signext %"#temp#.0")
  store %jl_value_t addrspace(10)* %29, %jl_value_t addrspace(10)** %12, align 8
  store %jl_value_t addrspace(10)* %25, %jl_value_t addrspace(10)** %.sub, align 8
  store %jl_value_t addrspace(10)* %29, %jl_value_t addrspace(10)** %14, align 8
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139705131220400 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %15, align 8
  %30 = call nonnull %jl_value_t addrspace(10)* @jl_f_getfield(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %.sub, i32 3)
;}
  %31 = bitcast %jl_value_t addrspace(10)* %30 to i64 addrspace(10)*
  %32 = getelementptr i64, i64 addrspace(10)* %31, i64 -1
  %33 = load i64, i64 addrspace(10)* %32, align 8
  %34 = and i64 %33, -16
  %35 = inttoptr i64 %34 to %jl_value_t*
  %36 = addrspacecast %jl_value_t* %35 to %jl_value_t addrspace(10)*
  %37 = icmp eq %jl_value_t addrspace(10)* %36, addrspacecast (%jl_value_t* inttoptr (i64 139705219414000 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %37, label %if2, label %L39

L66:                                              ; preds = %L57, %L2
  %merge = phi i8 [ 1, %L2 ], [ 0, %L57 ]
; Location: REPL[733]:5
  %38 = load i64, i64* %9, align 8
  store i64 %38, i64* %7, align 8
  ret i8 %merge

if2:                                              ; preds = %if
; Location: REPL[733]:3
  %39 = bitcast %jl_value_t addrspace(10)* %24 to i64 addrspace(10)*
  %40 = getelementptr i64, i64 addrspace(10)* %39, i64 -1
  %41 = load i64, i64 addrspace(10)* %40, align 8
  %42 = and i64 %41, -16
  %43 = inttoptr i64 %42 to %jl_value_t*
  %44 = addrspacecast %jl_value_t* %43 to %jl_value_t addrspace(10)*
  %45 = icmp eq %jl_value_t addrspace(10)* %44, addrspacecast (%jl_value_t* inttoptr (i64 139705219414000 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %45, label %if3, label %L32

L39:                                              ; preds = %if
  %46 = icmp eq %jl_value_t addrspace(10)* %36, addrspacecast (%jl_value_t* inttoptr (i64 139705133795312 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %46, label %if6, label %L55

if3:                                              ; preds = %if2
  %47 = bitcast %jl_value_t addrspace(10)* %24 to double addrspace(10)*
  %48 = load double, double addrspace(10)* %47, align 8
  %49 = bitcast %jl_value_t addrspace(10)* %30 to double addrspace(10)*
  %50 = load double, double addrspace(10)* %49, align 8
  %51 = call i8 @jlsys_isequal_30715(double %48, double %50)
  br label %L57

L32:                                              ; preds = %if2
  %52 = icmp eq %jl_value_t addrspace(10)* %44, addrspacecast (%jl_value_t* inttoptr (i64 139705133795312 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %52, label %if5, label %L55

L57:                                              ; preds = %if8, %if7, %if5, %if3
  %"#temp#1.0" = phi i8 [ %75, %if8 ], [ %71, %if7 ], [ %59, %if5 ], [ %51, %if3 ]
; Location: REPL[733]:2
; Function next; {
; Location: range.jl:471
; Function +; {
; Location: int.jl:53
  %53 = add nuw nsw i64 %"#temp#.0", 1
;}}
; Location: REPL[733]:3
  %54 = and i8 %"#temp#1.0", 1
  %55 = icmp eq i8 %54, 0
  br i1 %55, label %L66, label %L2

if5:                                              ; preds = %L32
  %56 = load i64, i64 addrspace(10)* %39, align 8
  %57 = bitcast %jl_value_t addrspace(10)* %30 to double addrspace(10)*
  %58 = load double, double addrspace(10)* %57, align 8
  %59 = call i8 @jlsys_isequal_42024(i64 %56, double %58)
  br label %L57

L55:                                              ; preds = %L46, %L39, %L32
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139705250501616 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %.sub, align 8
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 139705220669552 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %14, align 8
  %60 = call nonnull %jl_value_t addrspace(10)* @jl_apply_generic(%jl_value_t addrspace(10)** %.sub, i32 2)
  call void @llvm.trap()
  unreachable

if6:                                              ; preds = %L39
  %61 = bitcast %jl_value_t addrspace(10)* %24 to i64 addrspace(10)*
  %62 = getelementptr i64, i64 addrspace(10)* %61, i64 -1
  %63 = load i64, i64 addrspace(10)* %62, align 8
  %64 = and i64 %63, -16
  %65 = inttoptr i64 %64 to %jl_value_t*
  %66 = addrspacecast %jl_value_t* %65 to %jl_value_t addrspace(10)*
  %67 = icmp eq %jl_value_t addrspace(10)* %66, addrspacecast (%jl_value_t* inttoptr (i64 139705219414000 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %67, label %if7, label %L46

if7:                                              ; preds = %if6
  %68 = bitcast %jl_value_t addrspace(10)* %24 to double addrspace(10)*
  %69 = load double, double addrspace(10)* %68, align 8
  %70 = load i64, i64 addrspace(10)* %31, align 8
  %71 = call i8 @jlsys_isequal_42022(double %69, i64 %70)
  br label %L57

L46:                                              ; preds = %if6
  %72 = icmp eq %jl_value_t addrspace(10)* %66, addrspacecast (%jl_value_t* inttoptr (i64 139705133795312 to %jl_value_t*) to %jl_value_t addrspace(10)*)
  br i1 %72, label %if8, label %L55

if8:                                              ; preds = %L46
  %73 = load i64, i64 addrspace(10)* %61, align 8
  %74 = load i64, i64 addrspace(10)* %31, align 8
  %75 = call i8 @jlsys_isequal_44322(i64 %73, i64 %74)
  br label %L57
}

```

andyferris commented:
OK, thanks @Jutho, that makes a lot of sense. This PR was really aimed at `isless` (and I figured `isequal` should just follow the same pattern). For `n >= 15` the fully unrolled generated code is faster than the loop for heterogenous tuples.

I think the *correct* thing here is for me to update `isless` (and probably `<`) to the pattern Jutho made for `isequal` and `==`. We can discuss changing `Any16` to e.g. `Any32`, and an associated change to `MAX_TUPLETYPE_LEN`, elsewhere.

Commit:
Faster isless, <, and hash on Tuples

Uses same pattern as isequal and == so that a loop is used for large
tuples, and unrolled code (via recursion) for small tuples.

andyferris commented:
Alright, I've updated the PR to use the same method as `isequal` for `isless`, `<` and `hash`. Small tupples are unrolled via recursion and large tuples use a loop aglorithm.

Thanks everyone for you comments and suggestions (I made this PR an RFC for this reason) :smile:

vtjnash reviewed (approved):
Probably just needs a few tests now for the Any16 code

andyferris commented:
There are now tests for all the `Any16` code (including `isequal` and `==`).

If CI passes, I think we're ready to merge.

andyferris commented:
Appveyor and CircleCI seem to be failing everywhere. I'll merge.

nalimilan commented:
Sorry to be late, but it would have been nice to add tests including `missing` with long tuples. Else some of the new code won't be covered for 3VL. See examples at https://github.com/JuliaLang/julia/blob/aed88141e78862e791c76c5eb2ac3e965d9c6678/test/missing.jl#L264-L279
